iota-trinary/src/trytes.rs

```diff
@@ -0,0 +1,675 @@
+    /// the returned `Trytes` contains bytes `[at, len)`.
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `at` is beyond the last code point of the trytes.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::try_from("IOTA").unwrap();
+    /// let t2 = t.split_off(2);
+    /// assert_eq!(t.as_str(), "IO");
+    /// assert_eq!(t2.as_str(), "TA");
+    /// ```
+    pub fn split_off(&mut self, at: usize) -> Self {
+        let other = self.0.split_off(at);
+        Trytes(other)
+    }
+
+    /// Truncates this `Trytes`, removing all contents.
+    ///
+    /// While this means the `Trytes` will have a length of zero, it does not
+    /// touch its capacity.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::try_from("FOO").unwrap();
+    ///
+    /// t.clear();
+    ///
+    /// assert!(t.is_empty());
+    /// assert_eq!(0, t.len());
+    /// ```
+    pub fn clear(&mut self) {
+        self.0.clear()
+    }
+
+    /// Encode a string literal as bytes to trytes.
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::new();
+    ///
+    /// t.encode("Lorem");
+    ///
+    /// assert_eq!("VBCDFDTCAD", t.as_str());
+    /// ```
+    pub fn encode(&mut self, plain: &str) {
+        for byte in plain.bytes() {
+            let first = byte % 27;
+            let second = (byte - first) / 27;
+            self.0.put(TRYTE_ALPHABET[first as usize]);
+            self.0.put(TRYTE_ALPHABET[second as usize]);
+        }
+    }
+
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let t = Trytes::try_from("VBCDFDTCAD").unwrap();
+    ///
+    /// assert_eq!("Lorem", t.decode());
+    /// ```
+    pub fn decode(&self) -> String {
+        let mut s = String::with_capacity(self.len() / 2);
+        for slice in self.chunks(2) {
+            let first: u8 = TRYTE_ALPHABET.iter().position(|&x| x == slice[0]).unwrap() as u8;
+            let second: u8 = TRYTE_ALPHABET.iter().position(|&x| x == slice[1]).unwrap() as u8;
+            let decimal = first + second * 27;
+
+            s.push(decimal as char);
+        }
+        s
+    }
+
+    unsafe fn insert_bytes(&mut self, idx: usize, bytes: &[u8]) {
+        let len = self.len();
+        let amt = bytes.len();
+        self.0.reserve(amt);
+
+        ptr::copy(
+            self.0.as_ptr().add(idx),
+            self.0.as_mut_ptr().add(idx + amt),
+            len - idx,
+        );
+        ptr::copy(bytes.as_ptr(), self.0.as_mut_ptr().add(idx), amt);
+        self.0.set_len(len + amt);
+    }
+
+    fn all_tryte_alphabete<I>(vals: I) -> Result<()>
+    where
+        I: Iterator<Item = u8>,
+    {
+        let mut v = vals;
+        ensure!(v.all(Self::is_tryte_alphabete), "Invalid trytes alphabete.");
```

Comment:
I strongly recommend not using either `ensure!` or the `failure` crate in general. Whenever a failure occurs, the crate performs an allocation. This is likely to result in significantly slower code for a lot of use-cases. Instead, just roll your own error enum.